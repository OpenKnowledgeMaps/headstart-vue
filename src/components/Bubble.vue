<template>
    <g class="bubble_frame" :transform="translateString" :id="'bubble' + id">
        <circle
                :class="{ area: !zoomedOut, zoom_selected: zoomedOut }"
                :r="r"
                cx="0"
                cy="0"
                style="fill-opacity: 0.8; display: block;"
                @mouseover="$emit('bubbleMouseEnter', id)"
                @mouseout="$emit('bubbleMouseLeave', id)"
                @click="$emit('bubbleClick', id)"
                @dblclick.prevent="$emit('bubbleDoubleClick', id)"
        >
            <title>{{title}}</title>
        </circle>
        <foreignObject
                id="area_title_object"
                class="headstart"
                :x="foreignObjectX"
                :y="foreignObjectY"
                :width="inscribedSquareLength"
                :height="inscribedSquareLength"
                style="visibility: visible; display: block;"
                @mouseover="$emit('bubbleMouseEnter', id)"
                @mouseout="$emit('bubbleMouseLeave', id)"
                @click="$emit('bubbleClick', id)"
                @dblclick.prevent="$emit('bubbleDoubleClick', id)"
        >
            <body style="word-wrap: break-word;">
            <div id="area_title" :style="{width: inscribedSquareLength + 'px', height: inscribedSquareLength + 'px'}">
                <h2 class="highlightable" style="font-size: 12px;">
                    {{title}}
                </h2>
            </div>
            </body>
        </foreignObject>
    </g>
</template>

<script>
  export default {
    name: "Bubble",
    props: {
      id: Number,
      x: Number,
      y: Number,
      r: Number,
      title: String,
      zoomedOut: Boolean
    },
    computed: {
      translateString: function() {
        return "translate(" + this.x + "," + this.y + ")";
      },
      inscribedSquareLength: function () {
        return Math.sqrt(2*this.r*this.r);
      },
      foreignObjectX: function() {
        return -0.5*Math.sqrt(2*this.r*this.r);
      },
      foreignObjectY: function() {
        return -0.5*Math.sqrt(2*this.r*this.r);
      }
    }
  }
</script>