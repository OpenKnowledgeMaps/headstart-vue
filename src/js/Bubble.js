export default {
  name: 'Bubble',
  props: {
    id: Number,
    x: Number,
    y: Number,
    r: Number,
    order: Number,
    area: String,
    hovered: Boolean,
    zoomed: Boolean,
    selected: Boolean,
  },
  computed: {
    translateString() {
      return `translate(${this.x},${this.y})`;
    },
    inscribedSquareLength() {
      return Math.sqrt(2 * this.r * this.r);
    },
    foreignObjectX() {
      return -0.5 * Math.sqrt(2 * this.r * this.r);
    },
    foreignObjectY() {
      return -0.5 * Math.sqrt(2 * this.r * this.r);
    },
  },
};
