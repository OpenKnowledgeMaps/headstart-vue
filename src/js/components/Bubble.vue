<template src="../../templates/bubble.html"></template>

<script>
  export default {
    name: "Bubble",
    props: {
      id: Number,
      x: Number,
      y: Number,
      r: Number,
      order: Number,
      area: String,
      hovered: Boolean
    },
    computed: {
      translateString: function() {
        return "translate(" + this.x + "," + this.y + ")";
      },
      inscribedSquareLength: function () {
        return Math.sqrt(2*this.r*this.r);
      },
      foreignObjectX: function() {
        return -0.5*Math.sqrt(2*this.r*this.r);
      },
      foreignObjectY: function() {
        return -0.5*Math.sqrt(2*this.r*this.r);
      }
    }
  }
</script>
