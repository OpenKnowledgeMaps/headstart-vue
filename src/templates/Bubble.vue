<script src="../js/Bubble.js">
</script>

<template>
<g class="bubble_frame" :transform="translateString" :id="'bubble' + id">
    <circle
            :class="{ 
                area: (!hovered && !(zoomed && selected)), 
                zoom_selected: zoomed && selected}"
            style="fill-opacity: 0.8; display: block;"
            :r="r"
            cx="0"
            cy="0"
            @mouseover="$emit('mouseEnter', id)"
            @mouseout="$emit('mouseLeave', id)"
            @click="$emit('click', id)"
            @dblclick.prevent="$emit('doubleClick', id)"
    >
        <title>{{area}}</title>
    </circle>
    <foreignObject
            id="area_title_object"
            class="headstart"
            :x="foreignObjectX"
            :y="foreignObjectY"
            :width="inscribedSquareLength"
            :height="inscribedSquareLength"
            style="visibility: visible; display: block;"
            @mouseover="$emit('mouseEnter', id)"
            @mouseout="$emit('mouseLeave', id)"
            @click="$emit('click', id)"
            @dblclick.prevent="$emit('doubleClick', id)"
    >
        <body style="word-wrap: break-word;">
        <div id="areborder: 1px #000 solid;a_title" :style="{width: inscribedSquareLength + 'px', height: inscribedSquareLength + 'px'}">
            <h2 class="highlightable" style="font-size: 12px;" v-show="!hovered">
                {{area}}
            </h2>
        </div>
        </body>
    </foreignObject>
</g>
</template>