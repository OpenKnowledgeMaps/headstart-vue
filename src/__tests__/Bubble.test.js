import Bubble from '../components/Bubble';
import { mount } from '@vue/test-utils';

describe('Bubble', () => {
  let wrapper, circleWrapper, foreignObjectWrapper;
  beforeEach(() => {
    wrapper = mount(Bubble, {
      propsData: {
        id: 0,
        x: 100,
        y: 20,
        r: 30,
        title: "Bubble Title",
        zoomedOut: true
      }});
    circleWrapper = wrapper.find('circle');
    foreignObjectWrapper = wrapper.find('foreignObject');
  });

  it('renders a svg group element, a circle and a foreignobject according to its props', () => {
    expect(wrapper.find('g').exists()).toBe(true);
    expect(wrapper.find('circle').exists()).toBe(true);
    expect(wrapper.find('circle').attributes().r).toBe("30");
    expect(wrapper.find('#bubble0').exists()).toBe(true);
    expect(wrapper.find('#bubble0').attributes().transform).toBe("translate(100,20)");
    expect(wrapper.find('h2').text()).toBe("Bubble Title");
    expect(wrapper.find('foreignobject').exists()).toBe(true);
  });

  it('s circle element reacts to mouseover events', () => {
    circleWrapper.trigger('mouseover');
    expect(wrapper.emitted('bubbleMouseEnter')).toBeTruthy();
  });

  it('s circle element reacts to mouseout events', () => {
    circleWrapper.trigger('mouseout');
    expect(wrapper.emitted('bubbleMouseLeave')).toBeTruthy();
  });

  it('s circle element reacts to click events', () => {
    circleWrapper.trigger('click');
    expect(wrapper.emitted('bubbleClick')).toBeTruthy();
  });

  it('s circle element reacts to doubleclick events', () => {
    circleWrapper.trigger('dblclick');
    expect(wrapper.emitted('bubbleDoubleClick')).toBeTruthy();
  });

  it('s foreignobject element reacts to mouseover events', () => {
    foreignObjectWrapper.trigger('mouseover');
    expect(wrapper.emitted('bubbleMouseEnter')).toBeTruthy();
  });

  it('s foreignobject element reacts to mouseout events', () => {
    foreignObjectWrapper.trigger('mouseout');
    expect(wrapper.emitted('bubbleMouseLeave')).toBeTruthy();
  });

  it('s foreignobject element reacts to click events', () => {
    foreignObjectWrapper.trigger('click');
    expect(wrapper.emitted('bubbleClick')).toBeTruthy();
  });

  it('s foreignobject element reacts to doubleclick events', () => {
    foreignObjectWrapper.trigger('dblclick');
    expect(wrapper.emitted('bubbleDoubleClick')).toBeTruthy();
  });

  it('s title is not shown when zoomed out flag is set', () => {
    expect(wrapper.find('h2').isVisible()).toBe(false);
    wrapper.setProps({zoomedOut: false});
    expect(wrapper.find('h2').isVisible()).toBe(true);
  });

  //Snapshot Test!
  it('matches a previously made snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});