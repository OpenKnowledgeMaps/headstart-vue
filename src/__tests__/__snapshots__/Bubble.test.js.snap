// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bubble matches a previously made snapshot 1`] = `
<g transform="translate(300,300)" id="bubble1" class="bubble_frame">
  <circle r="74.4123" cx="0" cy="0" class="area" style="fill-opacity: 0.8; display: block;">
    <title>Cool</title>
  </circle>
  <foreignObject id="area_title_object" x="-52.617441933687736" y="-52.617441933687736" width="105.23488386737547" height="105.23488386737547" class="headstart" style="visibility: visible; display: block;">

    <body style="word-wrap: break-word;">
      <div id="area_title" style="width: 105.23488386737547px; height: 105.23488386737547px;">
        <h2 class="highlightable" style="font-size: 12px;">
          Cool
        </h2>
      </div>
    </body>
  </foreignObject>
</g>
`;
